File: routers/exports.py
import json
from pathlib import Path

from fastapi import APIRouter, HTTPException
from fastapi.responses import Response

from services.database import get_transcription

router = APIRouter()


# ── Helpers ───────────────────────────────────────────────────────────────────
def _ts(secs: float, sep: str = ",") -> str:
    h  = int(secs // 3600)
    m  = int((secs % 3600) // 60)
    s  = int(secs % 60)
    ms = round((secs % 1) * 1000)
    return f"{h:02d}:{m:02d}:{s:02d}{sep}{ms:03d}"


def _srt(segments: list) -> str:
    parts = []
    for i, seg in enumerate(segments, 1):
        parts.append(
            f"{i}\n{_ts(seg['start'])} --> {_ts(seg['end'])}\n{seg['text'].strip()}\n"
        )
    return "\n".join(parts)


def _vtt(segments: list) -> str:
    lines = ["WEBVTT", ""]
    for i, seg in enumerate(segments, 1):
        lines += [str(i), f"{_ts(seg['start'], '.')} --> {_ts(seg['end'], '.')}", seg["text"].strip(), ""]
    return "\n".join(lines)


def _tsv(segments: list) -> str:
    rows = ["start\tend\ttext"]
    for seg in segments:
        rows.append(f"{seg['start']:.3f}\t{seg['end']:.3f}\t{seg['text'].strip()}")
    return "\n".join(rows)


# ── Endpoint ──────────────────────────────────────────────────────────────────
@router.get("/{id}/{fmt}")
async def export_transcription(id: str, fmt: str):
    item = get_transcription(id)
    if not item:
        raise HTTPException(status_code=404, detail="Not found")
    if item["status"] != "completed":
        raise HTTPException(status_code=409, detail="Transcription not complete")

    segments = json.loads(item["segments"]) if item.get("segments") else []
    base     = Path(item["original_name"]).stem

    match fmt:
        case "txt":
            content, mime, ext = item["transcript"], "text/plain", "txt"
        case "srt":
            content, mime, ext = _srt(segments), "text/srt", "srt"
        case "vtt":
            content, mime, ext = _vtt(segments), "text/vtt", "vtt"
        case "tsv":
            content, mime, ext = _tsv(segments), "text/tab-separated-values", "tsv"
        case "json":
            content = json.dumps({
                "text":     item["transcript"],
                "language": item["language"],
                "duration": item["duration"],
                "segments": segments,
                "words":    json.loads(item["words"]) if item.get("words") else [],
            }, indent=2)
            mime, ext = "application/json", "json"
        case _:
            raise HTTPException(status_code=400, detail=f"Unknown format: {fmt}")

    return Response(
        content=content,
        media_type=f"{mime}; charset=utf-8",
        headers={"Content-Disposition": f'attachment; filename="{base}.{ext}"'},
    )

File: routes/exports.js
import { Router } from 'express';
import { getTranscription } from '../services/database.js';

const router = Router();

function secondsToTimestamp(secs, separator = ',') {
  const h = Math.floor(secs / 3600);
  const m = Math.floor((secs % 3600) / 60);
  const s = Math.floor(secs % 60);
  const ms = Math.round((secs % 1) * 1000);
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}${separator}${String(ms).padStart(3, '0')}`;
}

function buildSrt(segments) {
  return segments.map((seg, i) => (
    `${i + 1}\n${secondsToTimestamp(seg.start)} --> ${secondsToTimestamp(seg.end)}\n${seg.text.trim()}\n`
  )).join('\n');
}

function buildVtt(segments) {
  const lines = ['WEBVTT', ''];
  segments.forEach((seg, i) => {
    lines.push(`${i + 1}`);
    lines.push(`${secondsToTimestamp(seg.start, '.')} --> ${secondsToTimestamp(seg.end, '.')}`);
    lines.push(seg.text.trim());
    lines.push('');
  });
  return lines.join('\n');
}

function buildTxt(transcript) {
  return transcript;
}

function buildTsv(segments) {
  const rows = [['start', 'end', 'text'].join('\t')];
  segments.forEach(seg => {
    rows.push([seg.start.toFixed(3), seg.end.toFixed(3), seg.text.trim()].join('\t'));
  });
  return rows.join('\n');
}

// GET /api/exports/:id/:format
router.get('/:id/:format', (req, res) => {
  const { id, format } = req.params;
  const item = getTranscription(id);

  if (!item) return res.status(404).json({ error: 'Not found' });
  if (item.status !== 'completed') return res.status(409).json({ error: 'Transcription not complete' });

  const segments = item.segments ? JSON.parse(item.segments) : [];
  const baseName = item.original_name.replace(/\.[^/.]+$/, '');

  let content, mimeType, ext;

  switch (format) {
    case 'txt':
      content = buildTxt(item.transcript);
      mimeType = 'text/plain';
      ext = 'txt';
      break;
    case 'srt':
      content = buildSrt(segments);
      mimeType = 'text/srt';
      ext = 'srt';
      break;
    case 'vtt':
      content = buildVtt(segments);
      mimeType = 'text/vtt';
      ext = 'vtt';
      break;
    case 'tsv':
      content = buildTsv(segments);
      mimeType = 'text/tab-separated-values';
      ext = 'tsv';
      break;
    case 'json':
      content = JSON.stringify({
        text: item.transcript,
        language: item.language,
        duration: item.duration,
        segments,
        words: item.words ? JSON.parse(item.words) : [],
      }, null, 2);
      mimeType = 'application/json';
      ext = 'json';
      break;
    default:
      return res.status(400).json({ error: `Unknown format: ${format}` });
  }

  res.setHeader('Content-Type', `${mimeType}; charset=utf-8`);
  res.setHeader('Content-Disposition', `attachment; filename="${baseName}.${ext}"`);
  res.send(content);
});

export default router;
